Mono.Debugging/Mono.Debugging.Client/DebuggerSession.cs

```diff
@@ -352,42 +353,32 @@ public void SetOutputOptions(OutputOptions options)
 			}
 		}
 
-		readonly Queue<Action> actionsQueue = new Queue<Action>();
+		readonly ConcurrentQueue<Action> actionsQueue = new ConcurrentQueue<Action>();
 		bool threadExecuting;
 
 		void Dispatch (Action action)
 		{
 			if (UseOperationThread) {
-				lock (actionsQueue) {
-					actionsQueue.Enqueue (action);
-					if (!threadExecuting) {
-						threadExecuting = true;
-						ThreadPool.QueueUserWorkItem (delegate {
-							while (true) {
-								Action actionToExecute = null;
-								lock (actionsQueue) {
-									if (actionsQueue.Count > 0) {
-										actionToExecute = actionsQueue.Dequeue ();
-									} else {
-										threadExecuting = false;
-										return;
-									}
-								}
-								lock (slock) {
-									try {
-										actionToExecute ();
-									} catch (Exception ex) {
-										HandleException (ex);
-									}
+				actionsQueue.Enqueue (action);
+				if (!threadExecuting) {
+					threadExecuting = true;
```

Comment:
This is now race condition prone.

Reply:
There are at least 3 operations at hand: Load + if + store

Reply:
yeah... I didn't think about this boolean set. Need to see why that is needed

Reply:
Interlocked.CompareExchange might be useful here, but there are 2 objects that need to be synchronized - the boolean and the ConcurrentQueue.

That means we need to either:
1. Loop over setting the boolean with InterlockedCompareExchange - as it can "lose" the set because something else set it first - and then trigger the enqueue.
2. Use a `ConcurrentExclusiveSchedulerPair` and queue up on the tasks on the ExclusiveScheduler - but from the looks of it it is still synchronously locking
3. Use BlockingCollection instead? But that might also synchronously block.